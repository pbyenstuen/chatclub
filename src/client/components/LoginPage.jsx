import React, { useState } from "react";
import { Link } from "react-router-dom";
import { useHistory } from "react-router";
import useSubmit from "./useSubmit";
import InputField from "./InputField";

const LoginPage = ({ api, updateUser }) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const history = useHistory();

    const { handleSubmit: handleLogin, submitting, error } = useSubmit(
        async () => {
            await api.auth.logIn({ username, password });
        },
        async () => {
            await updateUser();
            history.push("/")
        }
    );

    return (
        <div id="login-container">
            <header id="auth-header">
                <h2>SIGN IN</h2>
            </header>
            <form onSubmit={handleLogin}>
                {submitting && <h4>Please wait...</h4>}
                {error && <h4>{error.toString()}</h4>}
                <InputField
                    label={"Username"}
                    value={username}
                    onValueChange={setUsername}
                />
                <InputField
                    label={"Password"}
                    type={"password"}
                    value={password}
                    onValueChange={setPassword}
                />
                <button className="submit-btn" disabled={submitting}>SIGN IN</button>
            </form>
            <div id="signup-options-container">
                <h3>Don't have an account?</h3>
                <Link to="/signup">SIGN UP</Link>
                <h4>OR</h4>
                <a href="/api/auth/google"><button id="google-btn">SIGN IN WITH GOOGLE</button></a>
            </div>
        </div>
    );
};

export default LoginPage;