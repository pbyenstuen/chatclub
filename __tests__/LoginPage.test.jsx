import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import LoginPage from "../src/client/components/LoginPage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("LoginPage", () => {
    it("can submit username and password", async () => {
        const logIn = jest.fn();
        const updateUser = jest.fn();

        const api = {
            auth: {
                logIn
            }
        }

        await mountComponent(<LoginPage api={api} updateUser={updateUser} />);

        Simulate.change(container.querySelector("input[name='username']"), {
            target: { value: "realDonaldTrump" },
        });

        Simulate.change(container.querySelector("input[name='password']"), {
            target: { value: "yourefired" },
        });

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(logIn).toBeCalledWith({
            username: "realDonaldTrump",
            password: "yourefired",
        });
    });

    it("can show error message", async () => {
        const api = {
            auth: {
                logIn: () => {
                    throw new Error("Invalid username/password");
                }
            }
        }

        await mountComponent(<LoginPage api={api} />);

        await act(async () => {
            await Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Error: Invalid username/password");
    });
});