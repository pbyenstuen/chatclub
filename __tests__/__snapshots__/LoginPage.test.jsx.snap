// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage can show error message 1`] = `"<div id=\\"login-container\\"><header id=\\"auth-header\\"><h2>SIGN IN</h2></header><form><h4>Error: Invalid username/password</h4><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"\\"></label></div><button class=\\"submit-btn\\">SIGN IN</button></form><div id=\\"signup-options-container\\"><h3>Don't have an account?</h3><a href=\\"/signup\\">SIGN UP</a><h4>OR</h4><a href=\\"/api/auth/google\\"><button id=\\"google-btn\\">SIGN IN WITH GOOGLE</button></a></div></div>"`;
